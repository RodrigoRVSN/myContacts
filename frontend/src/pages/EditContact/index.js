import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { PageHeader } from '../../components/PageHeader';

export function EditContact() {
  return (
    <>
      <PageHeader title="Editar Contato" />
      <ContactForm buttonLabel="Salvar" />
    </>
  );
}
