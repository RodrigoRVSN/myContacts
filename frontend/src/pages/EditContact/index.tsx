import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { Loader } from '../../components/Loader';
import { PageHeader } from '../../components/PageHeader';
import { useEditContact } from './useEditContact';

export default function EditContact() {
  const {
    isLoading, contactName, contactForm, handleSubmit,
  } = useEditContact();

  return (
    <>
      <Loader isLoading={isLoading} />

      <PageHeader title={isLoading ? 'Carregando...' : `Editar ${contactName}`} />

      <ContactForm buttonLabel="Salvar" ref={contactForm} onSubmit={handleSubmit} />
    </>
  );
}
