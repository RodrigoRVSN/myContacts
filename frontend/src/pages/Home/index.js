/* eslint-disable no-nested-ternary */
import React, {
  useCallback, useEffect, useMemo, useState,
} from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  Header,
  Card,
  InputSearchContainer,
  ListHeader,
  ErrorContainer,
  NoContactsContainer,
  NotFoundFilteredContacts,
} from './styles';

import arrow from '../../assets/images/icons/arrow.svg';
import edit from '../../assets/images/icons/edit.svg';
import trash from '../../assets/images/icons/trash.svg';
import sad from '../../assets/images/sad.svg';
import emptyBox from '../../assets/images/empty-box.svg';
import magnifierQuestion from '../../assets/images/magnifier-question.svg';

import { Loader } from '../../components/Loader';
import ContactsService from '../../services/ContactsService';
import Button from '../../components/Button';

export function Home() {
  const [contacts, setContacts] = useState([]);
  const [orderBy, setOrderBy] = useState('asc');
  const [searchTerm, setSearchTerm] = useState('');
  const [loading, setLoading] = useState(true);
  const [hasError, setError] = useState(false);

  const filteredContacts = useMemo(
    () =>
      contacts.filter((contact) =>
        contact.name.toLowerCase().includes(searchTerm.toLowerCase())),
    [contacts, searchTerm],
  );

  const loadContacts = useCallback(async () => {
    try {
      setLoading(true);
      const contactsList = await ContactsService.listContacts(orderBy);

      setError(false);
      setContacts(contactsList);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  }, [orderBy]);

  useEffect(() => {
    loadContacts();
  }, [loadContacts]);

  function handleToggleOrderBy() {
    setOrderBy((prevState) => (prevState === 'asc' ? 'desc' : 'asc'));
  }

  function handleChangeSearchTerm(ev) {
    setSearchTerm(ev.target.value);
  }

  function handleTryAgain() {
    loadContacts();
  }

  return (
    <>
      <Container>
        <Loader isLoading={loading} />

        {contacts.length > 0 && (
          <InputSearchContainer>
            <input
              value={searchTerm}
              onChange={(ev) => handleChangeSearchTerm(ev)}
              type="text"
              placeholder="Procure pelo contato"
            />
          </InputSearchContainer>
        )}

        <Header justifyContent={
          hasError
            ? 'flex-end'
            : (contacts.length > 0
              ? 'space-between'
              : 'center')
        }
        >
          {(!hasError && contacts.length > 0) && (
            <strong>
              {filteredContacts.length > 0
                ? `${filteredContacts.length} ${filteredContacts.length > 1 ? ' contatos' : ' contato'
                }`
                : 'Nenhum contato!'}
            </strong>
          )}
          <Link to="/new">Novo Contato</Link>
        </Header>

        {hasError && (
          <ErrorContainer>
            <img src={sad} alt="Sad" />
            <div className="error_container">
              <strong>Ocorreu um erro ao obter os seus contatos!</strong>
              <Button onClick={handleTryAgain}>Tentar novamente</Button>
            </div>
          </ErrorContainer>
        )}

        {!hasError && (
          <>
            {contacts.length < 1 && (
              <NoContactsContainer>
                <img src={emptyBox} alt="Sem contatos" />
                <p>
                  Você ainda não tem nenhum contato cadastrado!
                  Clique no botão <strong>”Novo contato”</strong> à cima
                  para cadastrar o seu primeiro!
                </p>
              </NoContactsContainer>
            )}

            {filteredContacts.length > 0 && (
              <ListHeader orderBy={orderBy}>
                <button onClick={handleToggleOrderBy} type="button">
                  <span>Nome</span>
                  <img src={arrow} alt="Arrow" />
                </button>
              </ListHeader>
            )}

            {(contacts.length > 0 && filteredContacts.length < 1)
              && (
                <NotFoundFilteredContacts>
                  <img src={magnifierQuestion} alt="magnifierQuestion" />
                  <span>
                    Nenhum resultado foi encontrado para <b>{searchTerm}</b>.
                  </span>
                </NotFoundFilteredContacts>
              )}

            {filteredContacts.map((contact) => (
              <Card key={contact.id}>
                <div className="info">
                  <div className="contact-name">
                    <strong>{contact.name}</strong>
                    {contact.category_name && (
                      <small>{contact.category_name}</small>
                    )}
                  </div>
                  <span>{contact.email}</span>
                  <span>{contact.phone}</span>
                </div>
                <div className="actions">
                  <Link to={`/edit/${contact.id}`}>
                    <img src={edit} alt="edit" />
                  </Link>
                  <button type="button">
                    <img src={trash} alt="trash" />
                  </button>
                </div>
              </Card>
            ))}
          </>
        )}

      </Container>
    </>
  );
}
