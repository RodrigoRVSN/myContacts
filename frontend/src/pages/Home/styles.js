import styled from 'styled-components';

export const Container = styled.div`
  margin-top: 2rem;
  position: relative;
`;

export const Header = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 2rem;

  strong{
    color: ${({ theme }) => theme.colors?.primary.lighter};
    font-size: 1.5rem;
  }

  a{
    color: ${({ theme }) => theme.colors?.primary.main};
    text-decoration: none;
    font-weight: bold;
    border: 2px solid ${({ theme }) => theme.colors?.primary.main};
    padding: 0.5rem 1rem;
    border-radius: 0.25rem;
    transition: all 0.2s ease-in;

    &:hover{
      background-color: ${({ theme }) => theme.colors?.primary.main};
      color: ${({ theme }) => theme.colors?.primary.light};
    }
  }
`;

export const ListHeader = styled.header`
  margin-top: 2rem;
  margin-bottom: 0.5rem;

    button{
      background: transparent;
      border: none;
      display: flex;
      align-items: center;

      span{
        margin-right: 0.5rem;
        font-weight: bold;
        color: ${({ theme }) => theme.colors?.primary.light};
      }

      img{
        transition: transform 0.2s ease-in;
        transform: ${({ orderBy }) => (orderBy === 'asc' ? 'rotate(180deg)' : 'rotate(0deg)')};
      }
    }
`;

export const Card = styled.div`
  background: ${({ theme }) => theme.colors?.primary.dark};
  color: ${({ theme }) => theme.colors?.primary.light};
  box-shadow: 0px 0.25rem 0.6rem rgba(0, 0, 0, 0.04);
  padding: 1rem;
  border-radius: 0.25rem;
  display: flex;
  align-items: center;
  justify-content: space-between;

  & + & {
    margin-top: 1rem;
  }

  .info{
    .contact-name{
    display: flex;
    align-items: center;

    small{
      color: ${({ theme }) => theme.colors?.primary.lighter};
      background: ${({ theme }) => theme.colors?.primary.main};
      font-weight: bold;
      text-transform: uppercase;
      padding: 0.25rem;
      border-radius: 0.25rem;
      margin-left: 0.5rem;
    }
  }
  span{
    display: block;
    font-size: 0.85rem;
    margin-top: 0.5rem;
    color: ${({ theme }) => theme.colors?.primary.lighter};
  }
}
  .actions{
  display: flex;
  align-items: center;

  button{
    background: transparent;
    border: none;
    margin-left: 0.5rem;
  }
}
`;

export const InputSearchContainer = styled.header`
width: 100%;
input{
  width: 100%;
  background: ${({ theme }) => theme.colors?.primary.dark};
  border: none;
  border-radius: 1.56rem;
  height: 3.2rem;
  box-shadow: 0 4px 10px rgba(0, 0, 0, 0.2);
  outline: 0;
  padding: 0 1rem;
  color: ${({ theme }) => theme.colors?.primary.lighter};
    &::placeholder{
    color: ${({ theme }) => theme.colors?.primary.lighter};
  }
}
`;
