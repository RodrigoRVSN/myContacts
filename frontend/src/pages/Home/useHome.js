import {
  useCallback, useEffect, useMemo, useState,
} from 'react';

import ContactsService from '../../services/ContactsService';
import { toast } from '../../utils/toast';

export const useHome = () => {
  const [contacts, setContacts] = useState([]);
  const [orderBy, setOrderBy] = useState('asc');
  const [searchTerm, setSearchTerm] = useState('');
  const [loading, setLoading] = useState(true);
  const [hasError, setError] = useState(false);
  const [isDeleteModalVisible, setIsModalVisible] = useState(false);
  const [contactBeingDeleted, setContactBeingDeleted] = useState(null);
  const [isLoadingDelete, setIsLoadingDelete] = useState(false);

  const filteredContacts = useMemo(
    () =>
      contacts.filter((contact) =>
        contact.name.toLowerCase().includes(searchTerm.toLowerCase())),
    [contacts, searchTerm],
  );

  const loadContacts = useCallback(async () => {
    try {
      setLoading(true);
      const contactsList = await ContactsService.listContacts(orderBy);

      setError(false);
      setContacts(contactsList);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  }, [orderBy]);

  useEffect(() => {
    loadContacts();
  }, [loadContacts]);

  function handleToggleOrderBy() {
    setOrderBy((prevState) => (prevState === 'asc' ? 'desc' : 'asc'));
  }

  function handleChangeSearchTerm(ev) {
    setSearchTerm(ev.target.value);
  }

  function handleTryAgain() {
    loadContacts();
  }

  function handleToggleDeleteModal() {
    setIsModalVisible((prevState) => !prevState);
    setContactBeingDeleted(null);
  }

  function handleDeleteContact(contact) {
    handleToggleDeleteModal();
    setContactBeingDeleted(contact);
  }

  async function handleConfirmDeleteContact() {
    try {
      setIsLoadingDelete(true);
      await ContactsService.deleteContact(contactBeingDeleted.id);
      toast({ type: 'success', text: 'Deletado!!' });
      handleToggleDeleteModal();
      setContacts((prevState) => prevState.filter(
        (contact) => contact.id !== contactBeingDeleted.id,
      ));
    } catch {
      toast({ type: 'error', text: 'Eita crabumco, erro!!' });
    } finally {
      setIsLoadingDelete(false);
    }
  }

  return {
    loading,
    isDeleteModalVisible,
    contactBeingDeleted,
    handleToggleDeleteModal,
    handleConfirmDeleteContact,
    isLoadingDelete,
    contacts,
    searchTerm,
    handleChangeSearchTerm,
    hasError,
    filteredContacts,
    handleTryAgain,
    orderBy,
    handleToggleOrderBy,
    handleDeleteContact,
  };
};
