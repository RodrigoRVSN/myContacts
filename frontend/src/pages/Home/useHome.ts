import {
  useDeferredValue,
  useCallback, useEffect, useState, useMemo, ChangeEvent,
} from 'react';

import ContactsService from '../../services/ContactsService';
import { IContact } from '../../types/IContact';
import { toast } from '../../utils/toast';

export const useHome = () => {
  const [contacts, setContacts] = useState<IContact[]>([]);
  const [orderBy, setOrderBy] = useState<'asc' | 'desc'>('asc');
  const [searchTerm, setSearchTerm] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const [hasError, setError] = useState(false);
  const [isDeleteModalVisible, setIsModalVisible] = useState(false);
  const [contactBeingDeleted, setContactBeingDeleted] = useState<IContact | null>(null);
  const [isLoadingDelete, setIsLoadingDelete] = useState(false);

  const deferredSearchTerm = useDeferredValue(searchTerm);

  const filteredContacts = useMemo(
    () =>
      contacts.filter((contact) =>
        contact.name.toLowerCase().includes(deferredSearchTerm.toLowerCase())),
    [contacts, deferredSearchTerm],
  );

  const loadContacts = useCallback(async (signal?: AbortSignal) => {
    try {
      setIsLoading(true);
      const contactsList = await ContactsService.listContacts(orderBy, signal!);

      setError(false);
      setContacts(contactsList);
    } catch (error) {
      if (error instanceof DOMException && error.name === 'AbortError') return;

      setError(true);
      setContacts([]);
    } finally {
      setIsLoading(false);
    }
  }, [orderBy]);

  useEffect(() => {
    const controller = new AbortController();

    loadContacts(controller.signal);

    return () => {
      controller.abort();
    };
  }, [loadContacts]);

  const handleToggleOrderBy = useCallback(() => {
    setOrderBy((prevState) => (prevState === 'asc' ? 'desc' : 'asc'));
  }, []);

  function handleChangeSearchTerm(ev: ChangeEvent<HTMLInputElement>) {
    setSearchTerm(ev.target.value);
  }

  function handleTryAgain() {
    loadContacts();
  }

  function handleToggleDeleteModal() {
    setIsModalVisible((prevState) => !prevState);
    setContactBeingDeleted(null);
  }

  const handleDeleteContact = useCallback((contact: IContact) => {
    handleToggleDeleteModal();
    setContactBeingDeleted(contact);
  }, []);

  async function handleConfirmDeleteContact() {
    try {
      setIsLoadingDelete(true);
      await ContactsService.deleteContact(contactBeingDeleted!.id);

      toast({ type: 'success', text: 'Deletado!!' });
      handleToggleDeleteModal();

      setContacts((prevState) => prevState.filter(
        (contact) => contact.id !== contactBeingDeleted!.id,
      ));
    } catch {
      toast({ type: 'error', text: 'Eita crabumco, erro!!' });
    } finally {
      setIsLoadingDelete(false);
    }
  }

  return {
    isLoading,
    isDeleteModalVisible,
    contactBeingDeleted,
    handleToggleDeleteModal,
    handleConfirmDeleteContact,
    isLoadingDelete,
    contacts,
    searchTerm,
    handleChangeSearchTerm,
    hasError,
    filteredContacts,
    handleTryAgain,
    orderBy,
    handleToggleOrderBy,
    handleDeleteContact,
  };
};
