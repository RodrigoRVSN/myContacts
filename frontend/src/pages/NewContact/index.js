import React from 'react';
import { PageHeader } from '../../components/PageHeader';
import { ContactForm } from '../../components/ContactForm';
import { useNewContact } from './useNewContact';

export function NewContact() {
  const { contactFormRef, handleSubmit } = useNewContact();

  return (
    <>
      <PageHeader title="Novo contato" />

      <ContactForm
        buttonLabel="Cadastrar"
        onSubmit={handleSubmit}
        ref={contactFormRef}
      />
    </>
  );
}
