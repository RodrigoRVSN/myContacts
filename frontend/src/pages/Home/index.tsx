import React from 'react';
import {
  Container,
} from './styles';

import { Loader } from '../../components/Loader';
import { Modal } from '../../components/Modal';
import { useHome } from './useHome';
import { InputSearch } from './components/InputSearch';
import { Header } from './components/Header';
import { ErrorStatus } from './components/ErrorStatus';
import { EmptyList } from './components/EmptyList';
import { SearchNotFound } from './components/SearchNotFound';
import { ContactsList } from './components/ContactsList';

export function Home() {
  const {
    isLoading,
    isDeleteModalVisible,
    contactBeingDeleted,
    handleToggleDeleteModal,
    handleConfirmDeleteContact,
    isLoadingDelete,
    contacts,
    searchTerm,
    handleChangeSearchTerm,
    hasError,
    filteredContacts,
    handleTryAgain,
    orderBy,
    handleToggleOrderBy,
    handleDeleteContact,
  } = useHome();

  const hasContacts = contacts.length > 0;
  const isListEmpty = !hasError && (!isLoading && !hasContacts);
  const isSearchEmpty = !hasError && hasContacts && filteredContacts.length < 1;

  return (
    <Container>
      <Loader isLoading={isLoading} />

      {hasContacts && (
        <InputSearch
          value={searchTerm}
          onChange={handleChangeSearchTerm}
        />
      )}

      <Header
        hasError={hasError}
        qtyContacts={contacts.length}
        qtyFilteredContacts={filteredContacts.length}
      />

      {hasError && <ErrorStatus onTryAgain={handleTryAgain} />}
      {isListEmpty && <EmptyList />}
      {isSearchEmpty && <SearchNotFound searchTerm={searchTerm} />}

      {hasContacts && (
      <>
        <ContactsList
          filteredContacts={filteredContacts}
          onDeleteContact={handleDeleteContact}
          orderBy={orderBy}
          onToggleOrderBy={handleToggleOrderBy}
        />

        <Modal
          danger
          isVisible={isDeleteModalVisible}
          title={`Tem certeza que deseja remover o contato "${contactBeingDeleted?.name}"?`}
          cancelLabel="Cancelar"
          confirmLabel="Confirmar"
          onCancel={handleToggleDeleteModal}
          onConfirm={handleConfirmDeleteContact}
          isLoading={isLoadingDelete}
        >
          <p>Esta ação não poderá ser desfeita!</p>
        </Modal>
      </>
      )}
    </Container>
  );
}
