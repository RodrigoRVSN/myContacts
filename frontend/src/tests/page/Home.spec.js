import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Home } from '../../pages/Home';

describe('<Home />', () => {
  it('should render home page', () => {
    const { container } = render(
      <Router>
        <Home />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
