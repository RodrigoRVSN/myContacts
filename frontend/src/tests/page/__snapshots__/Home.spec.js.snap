// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render home page 1`] = `
<div>
  <div
    class="sc-bdnxRM hZNSoo"
  >
    <header
      class="sc-eCApnc bdNUWk"
    >
      <input
        placeholder="Procure pelo contato"
        type="text"
      />
    </header>
    <div
      class="sc-gtsrHT lnOIuu"
    >
      <strong>
        3 contatos
      </strong>
      <a
        href="/new"
      >
        Novo Contato
      </a>
    </div>
    <div
      class="sc-dlnjwi uXPcy"
    >
      <header>
        <button
          type="button"
        >
          <span>
            Nome
          </span>
          <img
            alt="Arrow"
            src="arrow.svg"
          />
        </button>
      </header>
      <div
        class="sc-hKFxyN jjOrEm"
      >
        <div
          class="info"
        >
          <div
            class="contact-name"
          >
            <strong>
              Rodrigo Victor
            </strong>
            <small>
              Instagram
            </small>
          </div>
          <span>
            rodrigovictor81@gmail.com
          </span>
          <span>
            (15) 99715-4489
          </span>
        </div>
        <div
          class="actions"
        >
          <a
            href="/edit/123"
          >
            <img
              alt="edit"
              src="edit.svg"
            />
          </a>
          <button
            type="button"
          >
            <img
              alt="trash"
              src="trash.svg"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
